File: crates/defi-protocols/src/derivatives/generic_derivatives.rs
use async_trait::async_trait;

use defi_core::error::{DefiError, Result};
use defi_core::traits::Derivatives;
use defi_core::types::*;

pub struct GenericDerivatives {
    name: String,
    interface: String,
}

impl GenericDerivatives {
    pub fn from_entry(name: String, interface: String) -> Self {
        Self { name, interface }
    }
}

#[async_trait]
impl Derivatives for GenericDerivatives {
    fn name(&self) -> &str {
        &self.name
    }

    async fn build_open_position(&self, _params: DerivativesPositionParams) -> Result<DeFiTx> {
        Err(DefiError::Unsupported(format!(
            "[{}] Derivatives interface '{}' requires a protocol-specific adapter. \
            Supported: 'hlp_vault' (HLP Vault). Protocols like Rumpel need custom position management logic.",
            self.name, self.interface
        )))
    }

    async fn build_close_position(&self, _params: DerivativesPositionParams) -> Result<DeFiTx> {
        Err(DefiError::Unsupported(format!(
            "[{}] Derivatives interface '{}' requires a protocol-specific adapter. \
            Supported: 'hlp_vault' (HLP Vault). Protocols like Rumpel need custom position management logic.",
            self.name, self.interface
        )))
    }
}

File: ts/packages/defi-protocols/src/derivatives/generic_derivatives.ts
import type { IDerivatives } from "@hypurrquant/defi-core";
import {
  DefiError,
  type ProtocolEntry,
  type DerivativesPositionParams,
  type DeFiTx,
} from "@hypurrquant/defi-core";

export class GenericDerivativesAdapter implements IDerivatives {
  private readonly protocolName: string;
  private readonly interfaceName: string;

  constructor(entry: ProtocolEntry, _rpcUrl?: string) {
    this.protocolName = entry.name;
    this.interfaceName = entry.interface;
  }

  name(): string {
    return this.protocolName;
  }

  async buildOpenPosition(_params: DerivativesPositionParams): Promise<DeFiTx> {
    throw DefiError.unsupported(
      `[${this.protocolName}] Derivatives interface '${this.interfaceName}' requires a protocol-specific adapter. ` +
      `Supported: 'hlp_vault' (HLP Vault). Protocols like Rumpel need custom position management logic.`,
    );
  }

  async buildClosePosition(_params: DerivativesPositionParams): Promise<DeFiTx> {
    throw DefiError.unsupported(
      `[${this.protocolName}] Derivatives interface '${this.interfaceName}' requires a protocol-specific adapter. ` +
      `Supported: 'hlp_vault' (HLP Vault). Protocols like Rumpel need custom position management logic.`,
    );
  }
}
